### PR TITLE
Add load test for WebSub/WebSubHub

### PR DESCRIPTION
**Pull request by ayeshLK**
Fixes [#2174](https://github.com/ballerina-platform/ballerina-standard-library/issues/2174)

## Examples
N/A

## Checklist
- [X] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

File: load-tests/in_memory_hub/src/start_hub.bal [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import ballerina/websubhub;
+import in_memory_hub.dispatcher;
+
+public function main() returns error? {
+    // Initialize the Hub
+    _ = @strand {thread: "any"} start dispatcher:syncDispatcherState();
+
+    websubhub:Listener hubListener = check new (9090);
+    check hubListener.attach(hubService, "hub");
+    check hubListener.'start();
```
Comment (dilanSachi) [inline load-tests/in_memory_hub/src/start_hub.bal:26]: dont we need to attach the listener to the runtime?

Comment (ayeshLK) [reply load-tests/in_memory_hub/src/start_hub.bal:26]: We need this until the load test runs. So, we do not need to attach the listener to runtime.

**Review by dilanSachi — APPROVED** [review]
LGTM

**Comment by xlight05** [comment]
Currently, we don't have a mechanism to expose ports in the jmeter pod. Lets have a discussion on this.

File: load-tests/in_memory_hub/deployment/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+resources:
+  - in_memory_hub.yaml
```
Comment (xlight05) [inline load-tests/in_memory_hub/deployment/kustomization.yaml:2]: Currently c2c doesn't support websubhub listener generation, therefore it does not generate required k8s services. We need to fix this from c2c cide. https://github.com/ballerina-platform/module-ballerina-c2c/issues/483
Till its fixed, We can generate the service using Kustomize. 
You can add following contents.

Kustomizations.yaml
```
resources:
  - in_memory_hub.yaml
  - service.yaml

patches:
  - path: deployment-patch.yaml
    target:
      group: apps
      version: v1
      kind: Deployment
      name: in-memory-hub-deployment
```

deployment-patch.yaml
```yaml
apiVersion: "apps/v1"
kind: Deployment
metadata:
  name: no-name
spec:
  template:
    metadata:
      labels:
        logs: "true"
    spec:
      containers:
        - name: "in-memory-hub-deployment"
          imagePullPolicy: Always
          ports:
            - containerPort: 9090
              name: "port-1-in-mem"
              protocol: "TCP"
```
service.yaml
```yaml
---
apiVersion: "v1"
kind: "Service"
metadata:
  labels:
    app: "in_memory_hub"
  name: "in-memory-hub-svc"
spec:
  ports:
    - name: "port-1-in-mem"
      port: 9090
      protocol: "TCP"
      targetPort: 9090
  selector:
    app: "in_memory_hub"
  type: "ClusterIP"

```

File: load-tests/in_memory_hub/results/summary.csv [hunk 3]
```diff
@@ -0,0 +1 @@
+ Label,# Samples,Average,Median,90% Line,95% Line,99% Line,Min,Max,Error %,Throughput,Received KB/sec,Std. Dev.,Date,Payload,Users
```
Comment (xlight05) [inline load-tests/in_memory_hub/results/summary.csv:1]: Add a new line

Comment (ayeshLK) [reply load-tests/in_memory_hub/results/summary.csv]: Fixed with [e8ceaab7949ce2d86e64285cdfc600e1970d3ade](https://github.com/ballerina-platform/module-ballerina-websubhub/pull/432/commits/e8ceaab7949ce2d86e64285cdfc600e1970d3ade)

File: load-tests/in_memory_hub/deployment/kustomization.yaml (hunk 2 above)
Comment (ayeshLK) [reply load-tests/in_memory_hub/deployment/kustomization.yaml:2]: Fixed with [057ee41615f5c62a7aa8ceb2b0e58f60866d582d](https://github.com/ballerina-platform/module-ballerina-websubhub/pull/432/commits/057ee41615f5c62a7aa8ceb2b0e58f60866d582d)

File: load-tests/in_memory_hub/results/summary.csv [hunk 4]
```diff
@@ -0,0 +1,2 @@
...
+
```
Comment (ayeshLK) [inline load-tests/in_memory_hub/results/summary.csv:2]:
```suggestion
 Label,# Samples,Average,Median,90% Line,95% Line,99% Line,Min,Max,Error %,Throughput,Received KB/sec,Std. Dev.,Date,Payload,Users  
 
```

File: load-tests/in_memory_hub/deployment/service.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  name: "in-memory-hub-svc"
+spec:
+  ports:
+    - name: "port-1-in-mem"
+      port: 9090
+      protocol: "TCP"
+      targetPort: 9090
+  selector:
+    app: "in_memory_hub"
+  type: "ClusterIP"
```
Comment (xlight05) [inline load-tests/in_memory_hub/deployment/service.yaml:16]: Lets add this to end of this file as well. Service for the jmeter pod.
---
```
apiVersion: "v1"
kind: "Service"
metadata:
  labels:
    app: "jmeter"
  name: "client-svc"
spec:
  ports:
    - name: "client-port"
      port: 9100
      protocol: "TCP"
      targetPort: 9100
  selector:
    app: "jmeter"
  type: "ClusterIP"
  ```

File: load-tests/in_memory_hub/results/summary.csv [hunk 6]
```diff
@@ -0,0 +1,2 @@
+ Label,# Samples,Average,Median,90% Line,95% Line,99% Line,Min,Max,Error %,Throughput,Received KB/sec,Std. Dev.,Date,Payload,Users  
+
```
Comment (xlight05) [inline load-tests/in_memory_hub/results/summary.csv:2]: for some reason we can still see no eof error.

File: load-tests/in_memory_hub/scripts/load_test/main.bal [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+isolated function retrieveReceivedCount() returns int {
+    lock {
+        return receivedCount;
+    }
+}
+
+websub:SubscriberService subscriberService = @websub:SubscriberServiceConfig {
+    target: ["http://in-memory-hub-svc.default.svc.cluster.local:9090/hub", "test"],
+    callback: "http://in-memory-hub-svc.default.svc.cluster.local:9100/sub",
```
[the remark this answers is not in the document]
Comment (ayeshLK) [reply load-tests/in_memory_hub/scripts/load_test/main.bal]: Fixed with [789a2d2b988146e2d1d9c523c2ad267b0606a535](https://github.com/ballerina-platform/module-ballerina-websubhub/pull/432/commits/789a2d2b988146e2d1d9c523c2ad267b0606a535)

File: load-tests/in_memory_hub/deployment/service.yaml (hunk 5 above)
Comment (ayeshLK) [reply load-tests/in_memory_hub/deployment/service.yaml:16]: Fixed with [789a2d2b988146e2d1d9c523c2ad267b0606a535](https://github.com/ballerina-platform/module-ballerina-websubhub/pull/432/commits/789a2d2b988146e2d1d9c523c2ad267b0606a535)

**Review by xlight05 — APPROVED** [review]

**Pull request merged**